src/midi/midi_alsa.cpp

```diff
@@ -25,19 +25,46 @@
 
 #include <string>
 #include <sstream>
+#include <functional>
 
+#include "logging.h"
+#include "programs.h"
 #include "string_utils.h"
 
 #define ADDR_DELIM ".:"
 
-#if (SND_LIB_MAJOR >= 1) || ((SND_LIB_MAJOR == 0) && (SND_LIB_MINOR >= 6))
-#define snd_seq_flush_output(x)           snd_seq_drain_output(x)
-#define snd_seq_set_client_group(x, name) /* nop */
-#define my_snd_seq_open(seqp)             snd_seq_open(seqp, "hw", SND_SEQ_OPEN_OUTPUT, 0)
-#else
-/* SND_SEQ_OPEN_OUT causes oops on early version of ALSA */
-#define my_snd_seq_open(seqp) snd_seq_open(seqp, SND_SEQ_OPEN)
-#endif
+using port_action_t = std::function<void(snd_seq_client_info_t *client_info,
+                                         snd_seq_port_info_t *port_info)>;
+
+void for_each_alsa_seq_port(port_action_t action)
+{
+	// We can't reuse the sequencer from midi handler, as the function might
+	// be called before that sequencer is created.
+	snd_seq_t *seq;
+	if (snd_seq_open(&seq, "hw", SND_SEQ_OPEN_OUTPUT, 0) != 0) {
+		LOG_MSG("ALSA: Error: Can't open MIDI sequencer");
+		return;
+	}
+
+	snd_seq_client_info_t *client_info;
+	snd_seq_client_info_malloc(&client_info);
+
+	snd_seq_port_info_t *port_info;
+	snd_seq_port_info_malloc(&port_info);
+
+	snd_seq_client_info_set_client(client_info, -1);
+	while (snd_seq_query_next_client(seq, client_info) >= 0) {
+		const int client_id = snd_seq_client_info_get_client(client_info);
+		snd_seq_port_info_set_client(port_info, client_id);
+		snd_seq_port_info_set_port(port_info, -1);
+		while (snd_seq_query_next_port(seq, port_info) >= 0)
+			action(client_info, port_info);
+	}
```

Comment:
I'm not sure how to do it, but apparently we can query the port-info against these bit-masks:

![2021-03-19_11-06](https://user-images.githubusercontent.com/1557255/111824258-32e75400-88a3-11eb-870e-0fd1010b0892.png)

Would be very interesting to filter on these and mention the type.  So if Munt, Timidity, and QSynth were all running, we could get output like:
```
alsa:
   14:0 - Midi Through - Midi Through Port 0
  128:0 - MT-32 player
   129:0 - General MIDI player
   130:0 - Roland GS player
```

And presumbly filtering on MIDI would also knock out these:
```
    0:0 - system - Timer
    0:1 - system - announce
```

Reply:
That's exactly the plan :) I did some testing already and it seems like filtering by `SND_SEQ_PORT_TYPE_SYNTHESIZER` ("Generates sound") is the correct way (but I'm not 100% sure yet).

We need to have this not only to knock-off kernel-level sequencers 0:0 and 0:1, but also to filter out other sequencers that are there only for providing backwards-compatibility (e.g. PipeWire - it appears as a new sequencer, but it only forwards the MIDI from one API to another if synths are using PulseAudio backend).

However, I wanted to implement filtering in PR3, as I want to have a single implementation for both listing and selecting default ALSA port to connect to, so changes will be non-trivial.

Reply:
Ahh ok! It will come in the next PR; nice!